File: mobile/src/domain/notifier.ts
/**
 * Fill an outbound customer message template's placeholders. Port of the
 * placeholder half of backend/notifier.py. Template SELECTION now lives in
 * template-match.ts (conditional templates); this module only renders a body.
 */

/** Values available to a message template. */
export interface MessageEntry {
  customer?: string;
  phone?: string;
  expiry_date?: string;
  opendate?: string;
  telecom?: string;
  agency?: string;
  plan?: string;
  model?: string;
  staff?: string;
  milestone_offset?: number;
}

/** Derived time-since-open values ({months}/{years}), computed by the caller
 *  from the opendate + today (kept out of MessageEntry as they aren't row data). */
export interface RenderExtras {
  months?: number;
  years?: number;
}

/**
 * Fill the message template. Missing placeholders degrade to '' (matches the
 * Python _Safe dict). Placeholders: {customer} {telecom} {model} {expiry}
 * {opendate} {when} {phone} {agency} {plan} {staff} {offset} {months} {years}.
 */
export function renderMessage(
  template: string,
  entry: MessageEntry,
  when: string,
  extra: RenderExtras = {},
): string {
  const values: Record<string, string> = {
    customer: entry.customer ?? "",
    phone: entry.phone ?? "",
    expiry: entry.expiry_date ?? "",
    opendate: entry.opendate ?? "",
    telecom: entry.telecom ?? "",
    agency: entry.agency ?? "",
    plan: entry.plan ?? "",
    model: entry.model ?? "",
    staff: entry.staff ?? "",
    offset: String(entry.milestone_offset ?? 0),
    when,
    months: extra.months == null ? "" : String(extra.months),
    years: extra.years == null ? "" : String(extra.years),
  };
  return String(template ?? "").replace(/\{(\w+)\}/g, (_m, key: string) =>
    Object.prototype.hasOwnProperty.call(values, key) ? values[key] : "",
  );
}

File: backend/notifier.py
"""Render the outbound customer message from the template.

This app sends a message TO the customer (their 2-year contract is expiring), so
there are no internal notification channels (toast / webhook) here. The actual
delivery transport (KDE Connect / QR / SMS) is wired separately; this module only
turns the template + a due row into the final text.
"""
from __future__ import annotations

from typing import Any


def render_message(template: str, entry: dict[str, Any], when: str) -> str:
    """Fill the message template. Missing placeholders degrade to ''."""
    class _Safe(dict):
        def __missing__(self, key: str) -> str:  # noqa: D401
            return ""

    values = _Safe(
        customer=entry.get("customer", ""),
        phone=entry.get("phone", ""),
        expiry=entry.get("expiry_date", ""),
        opendate=entry.get("opendate", ""),
        telecom=entry.get("telecom", ""),
        agency=entry.get("agency", ""),
        plan=entry.get("plan", ""),
        model=entry.get("model", ""),
        staff=entry.get("staff", ""),
        offset=entry.get("milestone_offset", 0),
        when=when,
    )
    try:
        return template.format_map(values)
    except (KeyError, IndexError, ValueError):
        return template
